### ISSUE TITLE
Better camera controls [feature request]

### ISSUE THREAD

**Issue by evouga** [libigl/libigl#1130]
Currently the camera can be controlled by directly editing various intrinsic variables in `viewer.core`. These intrinsics have unclear and undocumented relationship to each other and to the view matrix; higher-level user-friendly API for programmatic camera control (panning, rotating, zooming) and queries (for the camera location and look/up/left directions in world coordinates; unprojecting screen pixels to world rays, etc) would be helpful.

#### Check all that apply (change to `[x]`)
- [X] Windows
- [X] Mac OS X
- [X] Linux

See <https://libigl.github.io/CONTRIBUTING/#bugreport> for more tips.

**Comment by jdumas** [comment]
That's a reasonable request. Note that there are already some `unproject_*` function for the last feature you mention (cf. tutorial 708_Picking). Panning, zooming and rotation can also be adjusted easily by modifying the member variables `camera_translation`, `camera_zoom` and `trackball_angle`. Not sure they deserve an additional method for that. The only "missing" one would be to retrieve camera position in world space, and maybe look up/left or smth.

**Comment by evouga** [comment]
Yeah it’s possible existing fields would suffice with documentation, but I don’t understand, for example, how “translation”, “base_translation”, and “eye” are related. Which one is actually the position of the camera? Which one do I change to translate the camera?

**Comment by jdumas** [comment]
`camera_base_translation` is the default translation of the camera to fit a model into the view (when you call `align_camera_center()`), and is added to the `camera_translation` which is controlled by the user/UI when interacting with the mouse. You can see the actual [operation here](https://github.com/libigl/libigl/blob/daf7d0707c833a21e7488a6d98fb0e6bd5d37f47/include/igl/opengl/ViewerCore.cpp#L138). Think of the base translation + rotation as kind of a model matrix (although Alec doesn't want to call it that way, so conceptually it's lumped into a single `view` matrix). `camera_eye` is the position of the camera after user scaling/zoom/translation has been applied, and typically wouldn't need to be touched.

**Comment by alecjacobson** [comment]
camera_base_translation and camera_base_zoom are regrettable. We should
just have a view matrix and expose setters and getters for camera-like
control.

**Comment by jdumas** [comment]
Do you suggest to also change the internals then? How would you do it to allow both fitting a view to a non-unit size model, while allowing for ui interaction like zooming/panning. I'm also not sure whether a `set_camera_translation(...)` should only set `camera_base_translation`, or set `camera_translation`. It's kinda weird indeed...

**Comment by alecjacobson** [comment]
Yes, I think we should change the internals so that there's only a view
matrix. When loading a non-unit model the view matrix can be automatically
adjust to view it at full screen.
-Alec

>

**Comment by jdumas** [comment]
So then panning/zooming would be implemented by premultiplying the view matrix with a translation/scaling matrix? That sounds neat. We could also move this [update_matrices](https://github.com/libigl/libigl/blob/daf7d0707c833a21e7488a6d98fb0e6bd5d37f47/include/igl/opengl/ViewerCore.cpp#L128) block into its own method for example. But we'd still need to keep most of the `camera_eye`, `camera_dnear` and other such parameters to compute the initial `view` matrix.
